### PR TITLE
Oren/new presence

### PR DESCRIPTION
**Pull request by orenor10**

### CODE REVIEW

**Comment by assafshp** [comment]
jenkins-test

File: dist/presence/presence.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                    var userRef = rootRef.child(PRESENCE_PATH + userId);
+                    userRef.on('value', trackUserPresenceCB.bind(null, cb, userId));
+                };
+
+                PresenceService.stopTrackUserPresence = function (userId) {
+                    var userRef = rootRef.child(PRESENCE_PATH + userId);
+                    userRef.off('value', trackUserPresenceCB);
+                };
+
+                function trackUserPresenceCB(cb, userId, snapshot) {
```
Comment (assafshp) [inline dist/presence/presence.js]: are you sure you getting the last snapshot param...

userRef.on('value', trackUserPresenceCB.bind(null, cb, userId));

File: dist/znk.infra.js [hunk 2]
```diff
@@ -379,6 +418,57 @@ angular.module('znk.infra.analytics').run(['$templateCache', function($templateC
...
+
+    angular.module('znk.infra.assignModule').service('UserAssignModuleService', [
+        'ZnkModuleService', 'ModuleResultsService', '$q',
+        function (ZnkModuleService, ModuleResultsService, $q) {
+            var userAssignModuleService = {};
+
+            userAssignModuleService.getUserAssignModules = function (userId) {
+                return ModuleResultsService.getUserModuleResultsGuids(userId).then(function (resultsGuids) {
+                    var moduleResults = {};
+                    var getProm = $q.when();
```
Comment (assafshp) [inline dist/znk.infra.js]: why do you need this promise ?

File: dist/znk.infra.js [hunk 3]
```diff
@@ -379,6 +418,57 @@ angular.module('znk.infra.analytics').run(['$templateCache', function($templateC
...
+                    angular.forEach(resultsGuids, function (resultGuid) {
+                        getProm = getProm.then(function(){
```
Comment (assafshp) [inline dist/znk.infra.js]: why?

File: dist/znk.infra.js [hunk 4]
```diff
@@ -379,6 +418,57 @@ angular.module('znk.infra.analytics').run(['$templateCache', function($templateC
...
+                });
+            };
+
+            return userAssignModuleService;
+        }
+    ]);
+})(angular);
+
+
+angular.module('znk.infra.assignModule').run(['$templateCache', function($templateCache) {
```
Comment (assafshp) [inline dist/znk.infra.js]: are you sure it is needed?

File: test/spec/moduleResult/moduleResult.spec.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
+describe('testing service "ModuleResults":', function () {
+    'use strict';
+
+    //beforeEach(module('znk.infra.moduleResults', 'znk.infra.storage', 'znk.infra.enum', 'htmlTemplates', 'testUtility', 'storage.mock', 'user.mock'));
```
Comment (assafshp) [inline test/spec/moduleResult/moduleResult.spec.js]: delete if not needed

File: test/spec/moduleResult/moduleResult.spec.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    beforeEach(module('znk.infra.moduleResults', 'znk.infra.utility', 'znk.infra.config', 'znk.infra.storage', 'testUtility', 'storage.mock'));
+
+
+    var ModuleResultsService, UtilitySrv, actions, testStorage, InfraConfigSrv, $q, TEST_UID;
+    beforeEach(inject([
+        '$injector',
+        function ($injector) {
+            TEST_UID = '68639c20-f951-47a4-b4e7-de0c9142e39f';
+            var UtilitySrv = $injector.get('UtilitySrv');
```
Comment (assafshp) [inline test/spec/moduleResult/moduleResult.spec.js]: it is already defined by var in line 8

File: test/spec/moduleResult/moduleResult.spec.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            ModuleResultsService = $injector.get('ModuleResultsService');
```
Comment (assafshp) [inline test/spec/moduleResult/moduleResult.spec.js]: already defined in line 8

File: test/spec/moduleResult/moduleResult.spec.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    beforeEach(function () {
+        testStorage.db.moduleResults = {};
+        testStorage.db.users = {
+            '$$uid': {
+                moduleResults: {}
+            }
+        };
+    });
+
+    it('when requesting for a not exiting result then a new initialized result should be returned', function () {
```
Comment (assafshp) [inline test/spec/moduleResult/moduleResult.spec.js]: exiting ->typo existing

Comment (assafshp) [inline test/spec/moduleResult/moduleResult.spec.js]: good test

File: test/spec/moduleResult/moduleResult.spec.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        var moduleId = 1;
+        var exerciseResult = actions.getModuleResultById(moduleId, TEST_UID, true);
```
Comment (assafshp) [inline test/spec/moduleResult/moduleResult.spec.js]: small issue: exerciseResult is moduleResult so why to call it exercise

File: dist/moduleResults/moduleResults.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                var userResultsPath = USER_MODULE_RESULTS_PATH.replace('$$uid', userId);
+                return storage.get(userResultsPath);
+            };
+
+            moduleResultsService.getModuleResultByGuid = function (resultGuid, defaultValue) {
+                var resultPath = MODULE_RESULTS_PATH + '/' + resultGuid;
+                return storage.get(resultPath, defaultValue);
+            };
+
+            moduleResultsService.getModuleResultById = function (moduleId, userId, withDefaultResult) {
```
Comment (assafshp) [inline dist/moduleResults/moduleResults.js]: Do you mean by which ID? userId ? 

I prefer a better name - getModuleResultByUid (I didn't understand what is the diff between getModuleResultByGuid

File: test/spec/moduleResult/moduleResult.spec.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    it('when requesting for a not exiting result with dont initialize parameter then a new result should not be created and null should be returned', function () {
+        var moduleId = 1;
+        var exerciseResult = actions.getModuleResultById(moduleId, TEST_UID, false);
+        expect(exerciseResult).toBeNull();
+        expect(testStorage.db.moduleResults[moduleId]).toBeUndefined();
+    });
+
+    it('when requesting for a not exiting result By Guid with no default, null should return', function () {
+        var moduleId = 1;
+        var exerciseResult = actions.getModuleResultByGuid('123-456-789');
```
Comment (assafshp) [inline test/spec/moduleResult/moduleResult.spec.js]: '123-456-789' should be const

File: test/spec/moduleResult/moduleResult.spec.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            moduleId: moduleId,
+            tutorId: null,
+            assign: false,
+            contentAssign: false,
+            uid: TEST_UID
+        };
+        var exerciseResult = actions.getModuleResultByGuid('123-456-789', expectedModuleResult);
+        expect(exerciseResult).toEqual(expectedModuleResult);
+    });
+});
```
Comment (assafshp) [inline test/spec/moduleResult/moduleResult.spec.js]: what with a real test to see you getting a real object with data?

File: dist/znk.infra.js (hunk 2 above)
Comment (orenor10) [inline dist/znk.infra.js]: its a trick to make sure $q.all will wait all the promises we create within angular.forEach

File: dist/znk.infra.js (hunk 3 above)
Comment (orenor10) [inline dist/znk.infra.js]: the same reason as previous comment

File: dist/znk.infra.js (hunk 4 above)
Comment (orenor10) [inline dist/znk.infra.js]: this file is from dist folder. the build task add it

File: test/spec/moduleResult/moduleResult.spec.js (hunk 5 above)
Comment (orenor10) [inline test/spec/moduleResult/moduleResult.spec.js]: done

File: test/spec/moduleResult/moduleResult.spec.js (hunk 6 above)
Comment (orenor10) [inline test/spec/moduleResult/moduleResult.spec.js]: correct. fixed

File: test/spec/moduleResult/moduleResult.spec.js (hunk 7 above)
Comment (orenor10) [inline test/spec/moduleResult/moduleResult.spec.js]: no new definition made here

File: test/spec/moduleResult/moduleResult.spec.js (hunk 8 above)
Comment (orenor10) [inline test/spec/moduleResult/moduleResult.spec.js]: fixed

File: test/spec/moduleResult/moduleResult.spec.js (hunk 9 above)
Comment (orenor10) [inline test/spec/moduleResult/moduleResult.spec.js]: moduleResults is a better name. I changed it

File: dist/moduleResults/moduleResults.js (hunk 10 above)
Comment (orenor10) [inline dist/moduleResults/moduleResults.js]: 1. the function return result for specific module id. I canged the name to getModuleResultByModuleId
2. getModuleResultByGuid used for request getUserModuleResultsGuids (list of guids), then I used getModuleResultByGuid to get specific results.

getModuleResultByModuleId will be used in the student app

File: test/spec/moduleResult/moduleResult.spec.js (hunk 11 above)
Comment (orenor10) [inline test/spec/moduleResult/moduleResult.spec.js]: done

File: test/spec/moduleResult/moduleResult.spec.js (hunk 12 above)
Comment (orenor10) [inline test/spec/moduleResult/moduleResult.spec.js]: the 1st test made it...

File: src/components/presence/presence.service.js [hunk 13]
```diff
@@ -13,7 +13,7 @@
             '$log', '$injector', 'ENV', '$rootScope', 'storageFirebaseAdapter',
             function ($log, $injector, ENV, $rootScope, storageFirebaseAdapter) {
                 var PresenceService = {};
-                var rootRef = storageFirebaseAdapter.getRef();
+                var rootRef = storageFirebaseAdapter(ENV.fbDataEndPoint);
```
Comment (assafshp) [inline src/components/presence/presence.service.js]: ?

Comment (orenor10) [inline src/components/presence/presence.service.js]: the storageFirebaseAdapter return an object with getRef fuction within

**Pull request merged**